pywbem_mock/_namespaceprovider.py

```diff
@@ -0,0 +1,290 @@
+        except KeyError:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Namespace creation via CreateInstance: "
+                        "Missing 'Name' property in the {0!A} instance ",
+                        new_instance.classname))
+
+        # Normalize the namespace name
+        # TODO: Somehow we need to get this out of the way.
+        new_namespace = new_namespace.strip('/')
+
+        # Write it back to the instance in case it was changed
+        new_instance['Name'] = new_namespace
+
+        # These values must match those in
+        # tests/unittest/utils/wbemserver_mock.py.
+        try:
+            if new_instance['CreationClassName'] != new_instance.classname \
+                or new_instance['ObjectManagerName'] != OBJECTMANAGERNAME \
+                or new_instance['ObjectManagerCreationClassName'] != \
+                'CIM_ObjectManager' \
+                or new_instance['SystemName'] != SYSTEMNAME \
+                or new_instance['SystemCreationClassName'] != \
+                    SYSTEMCREATIONCLASSNAME:
+
+                raise CIMError(
+                    CIM_ERR_INVALID_PARAMETER,
+                    _format("Namespace creation via CreateInstance: "
+                            "Invalid property for CIM_Instance {0!A}",
+                            new_instance))
+
+        except KeyError as ke:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Namespace creation via CreateInstance: "
+                        "Missing property for CIM_Instance {0!A}. Exception "
+                        "{1!A}", new_instance, ke))
+
+        if new_namespace not in self.cimrepository.namespaces:
+            self.cimrepository.add_namespace(new_namespace)
+
+        # Pass NewInstance to the default CreateInstance
+        # Default CreateInstance creates path, validates all key properties
+        # in instance and inserts instance into the repository.
+        return super(CIMNamespaceProvider, self).CreateInstance(namespace,
+                                                                new_instance)
+
+    def ModifyInstance(self, ModifiedInstance,
+                       IncludeQualifiers=None, PropertyList=None):
+        """
+        Modification of CIM_Namespace instance not allowed
+        """
+        raise CIMError(
+            CIM_ERR_NOT_SUPPORTED,
+            _format("Modification of CIM_Namespace {0!A} not implemented. ",
+                    ModifiedInstance.classname))
+
+    def DeleteInstance(self, InstanceName):
+        """
+        Delete the namespace defined by the Name property in InstanceName
+        and remove the instance defined by InstanceName
+        """
+        remove_namespace = InstanceName.keybindings['Name']
+
+        # Reflect the namespace deletion in the CIM repository
+        # This will check the namespace for being empty.
+        self.remove_namespace(remove_namespace)
```

Comment:
In `CreateInstance()` the `add_namespace()` is invoked via `self.cimrepository`. Here, the wrapper function on the repository class is invoked, which adds CIM-level error handling. That is inconsistent.

We should think about that structure and whether it is a good idea to have wrapper functions that add only slight things (such as CIM-level error handling).

Reply:
TODO

Reply:
KARL Check this.

Reply:
We agreed to just invoke `remove_namespace()` through `self.cimrepository` and leave the other changes for a future PR.

Reply:
This is another case where one method calls another.  This method checks the namespace
parameter for none and then calls the CIM repository namespace rasining the CIMError
which is passed back to the user.

I did not change it.